src/math/bvh.js

```diff
@@ -0,0 +1,451 @@
+                        if (this.minDist == null) {
+                            this.minDist = dist;
+                        }
+                        this.minDist = Math.min(this.minDist, dist);
+                    }
+                }
+                if (stackPtr === 0) {
+                    break;
+                } else {
+                    node = stack[--stackPtr];
+                    continue;
+                }
+            }
+            let child1 = this.bvhNode[node.leftFirst];
+            let child2 = this.bvhNode[node.leftFirst + 1];
+            let dist1 = this.IntersectAABB(ray, child1.aabbMin, child1.aabbMax);
+            let dist2 = this.IntersectAABB(ray, child2.aabbMin, child2.aabbMax);
+            if (dist1 > dist2) {
+                [dist1, dist2] = [dist2, dist1];
+                [child1, child2] = [child2, child1];
+            }
+            if (dist1 === INFINITY) {
+                if (stackPtr === 0) {
+                    break;
+                } else {
+                    node = stack[--stackPtr];
+                }
+            } else {
+                node = child1;
+                if (dist2 !== INFINITY) {
+                    stack[stackPtr++] = child2;
+                }
+            }
+        }
+    }
+
+    /**
+     * Calculates whether a ray intersects an aabb and returns the distance between the ray origin and aabb if it does
+     *
+     * @param {Ray} [ray] - The ray to be traced
+     * @param {Vec3} [bmin] - The min corner of the aabb
+     * @param {Vec3} [bmax] - The max corner of the aabb
+     * @returns {number} The distance between the ray origin and the aabb, INFINITY if there is no intersection
+     * @ignore
+     */
+    IntersectAABB(ray, bmin, bmax) {
+        if (ray.rD == null) {
+            ray.rDx = 1 / ray.direction.x;
+            ray.rDy = 1 / ray.direction.y;
+            ray.rDz = 1 / ray.direction.z;
+        }
+        const tx1 = (bmin.x - ray.origin.x) * ray.rDx;
+        const tx2 = (bmax.x - ray.origin.x) * ray.rDx;
+        let tmin = Math.min(tx1, tx2);
+        let tmax = Math.max(tx1, tx2);
+        const ty1 = (bmin.y - ray.origin.y) * ray.rDy;
+        const ty2 = (bmax.y - ray.origin.y) * ray.rDy;
+        tmin = Math.max(tmin, Math.min(ty1, ty2));
+        tmax = Math.min(tmax, Math.max(ty1, ty2));
+        const tz1 = (bmin.z - ray.origin.z) * ray.rDz;
+        const tz2 = (bmax.z - ray.origin.z) * ray.rDz;
+        tmin = Math.max(tmin, Math.min(tz1, tz2));
+        tmax = Math.min(tmax, Math.max(tz1, tz2));
+        if (tmax >= tmin  && tmax > 0 && (this.minDist == null || (this.minDist != null && tmin < this.minDist))) {
+            return tmin;
+        }
+        return INFINITY;
+    }
+}
+
+function vec3MinInPlace(a, b, c) {
+    c.set(Math.min(a.x, b.x), Math.min(a.y, b.y), Math.min(a.z, b.z));
+    return c;
+}
+
+function vec3MaxInPlace(a, b, c) {
+    if (c == null) {
+        c = new Vec3();
+    }
```

Comment:
1) This looks strange, is this supposed to work as default argument?
2) Please add JSDoc
3) Prefer `===` over `==`